docs/proposals/runtime-image.md

```diff
@@ -0,0 +1,122 @@
+    base:
+      image: runtime-base-image:latest
+    workDir: /workspace/source
+    env:
+      JAVA_HOME: /path/to/java-home
+    labels:
+      custom-image-label: value
+    directories:
+      - /path/to/directory
+      - /path/to/another/directory:/target/location
+    entrypoint:
+      - java
+      - -jar
+      - ...
+```
+
+The new attributes added are:
+
+- `spec.runtime.base`: specifies the runtime base-image to be used, using [Image](https://github.com/redhat-developer/build/blob/97012ab56417ce1691a70896d90e490ea6a4d23c/pkg/apis/build/v1alpha1/build_types.go#L58) as type;
+- `spec.runtime.workDir`: path to `WORKDIR` in runtime-image;
+- `spec.runtime.env`: runtime-image additional environment variables, key-value;
+- `spec.runtime.labels`: runtime-image additional labels, key-value;
+- `spec.runtime.directories`: list of directories paths to be copied to runtime-image, those can be defined as `<source>:<destination>` split by colon (`:`);
+- `spec.runtime.entrypoint`: entrypoint command, specified as a list;
+
+## Runtime-Image
+
+Given the API changes described above, we will now address how the runtime-image will be created. This image is always based on an image built by the `BuildStrategy` (or `ClusterBuildStrategy`) steps, and therefore, we can introduce new steps to add the elements needed for the runtime-image creation.
+
+
+### Runtime Dockerfile
+
+The first step to be appended on the given `BuildStrategy` instance is the `Dockerfile` for the runtime-image, commonly named `Dockerfile.runtime`.
+
+This Dockerfile would like the following example:
+
+
+```
+FROM output-image:tag as builder
+
+FROM runtime-image:tag
+ENV VARIABLE=value
+LABEL label=value
+COPY --from=builder "/path/to/source" "/path/to/destination"
+WORKDIR "/path/to/destination"
+ENTRYPOINT [ "command", "args" ]
+```
+
+The example shows a [multi-stage image build](https://docs.docker.com/develop/develop-images/multistage-build/) using the outcomes of the original `BuildStrategy` steps as `builder` image, and afterwards, rendering all the elements supported by the [new `runtime` section](#API-Changes). To create this file we may employ [`text/template`](https://golang.org/pkg/text/template/) to render its contents.
+
+### Appending Steps
+
+During the [`BuildRun` resource reconciliation](https://github.com/redhat-developer/build/blob/97012ab56417ce1691a70896d90e490ea6a4d23c/pkg/controller/buildrun/buildrun_controller.go#L116) we have the opportunity to amend the `BuildStrategy` instance. The first step to be added is the `Dockerfile.runtime` creation:
+
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: BuildStrategy
+# ...
+spec:
+  buildSteps:
+    - name: step-runtime-dockerfile
+  	  image: $(build.builder.image)
+  	  securityContext:
+        runAsUser: 0
+  	  workingdir: /workspace/source
+  	  command:
+        - /bin/bash
+  	  args:
+        - -x
+        - -c
+        - >
+          echo '<DOCKERFILE_CONTENT>' >Dockerfile.runtime
+```
```

Comment:
My gut feeling is that I do not agree that this should be implemented like an amendment to the build strategy, even if only in memory. Instead, `generate_taskrun` should parse the runtime section of the build and act accordingly during taskrun creation.

Reply:
In other words, operator at a lower level of abstraction? Sounds OK to me!

Reply:
I'll explore the `generate_taskrun` route for this. Thanks for the input!

Reply:
Please consider the latest commits.